lib/std/testing.zig

```diff
@@ -374,6 +374,43 @@ pub const TmpDir = struct {
         self.parent_dir.close();
         self.* = undefined;
     }
+
+    /// Writes program string as zig file into tmp directory
+    /// Caller owns memory
+    ///
+    /// ```
+    /// const progstr = "pub fn main() void {}\n";
+    /// var it = try std.process.argsWithAllocator(std.testing.allocator);
+    /// defer it.deinit(); // no-op unless WASI or Windows
+    /// const testargs = try std.testing.getTestArgs(&it);
+    /// var tmp = std.testing.tmpDir(.{ .no_follow = true }); // ie zig-cache/tmp/8DLgoSEqz593PAEE
+    /// defer tmp.cleanup();
+    /// const zigfile_path = try tmp.writeZigFile(std.testing.allocator, progstr, "bruh");
+    /// defer std.testing.allocator.free(zigfile_path);
+    /// const binary = zigfile_path[0 .. zigfile_path.len - 4]; // '.zig' is 4 characters
+    /// try std.testing.buildExe(testargs.zigexec, zigfile_path, binary);
+    /// ```
+    pub fn writeZigFile(
+        self: *TmpDir,
+        alloc: std.mem.Allocator,
+        progstr: []const u8,
+        filename: []const u8,
+    ) ![]const u8 {
+        const tmpdir_path = try self.getFullPath(alloc);
+        defer alloc.free(tmpdir_path);
+        const suffix_zig = ".zig";
+        const zigfile_path = try std.mem.concat(alloc, u8, &[_][]const u8{
+            tmpdir_path,
+            std.fs.path.sep_str,
+            filename,
+            suffix_zig,
+        });
+        errdefer alloc.free(zigfile_path);
+        const zigfile = try std.mem.concat(alloc, u8, &[_][]const u8{ filename, suffix_zig });
+        defer alloc.free(zigfile);
+        try self.dir.writeFile(zigfile, progstr);
```

Comment:
This unfortunately ignores dir and uses an absolute path. Ideally instead it will write a relative name based on dir.

Reply:
The underlying problem is that `tmpDir(opts: std.fs.Dir.OpenDirOptions) TmpDir` creates dirs based on `cwd` with `cwd/cache/tmp/base64random()`, but `cwd` is not stored in `tmpDir`.

The 4 options are:
1. compute absolute path to compute relative based on `cwd`
    - 1.1 new method getRelativePath identical to getFullPath except for applying `std.fs.path.relative` to get relative path
    - 1.2 use getFullPath() + another `cwd` + `std.fs.path.relative`
2. store `cwd` in struct `TmpDir`
3. remove all the path stuff with comments that user should use `dir.createFile()/closeFile()`: slight/big annoyance here one is forced to use `/// Call `File.close` on the result when done.`
    - Biggest drawback: Users will likely reimplement the "getmepath" functions for debugging stuff.

Reply:
Nevermind actually. Only thing I need is
```zig
        const relative_path = try fs.path.join(allocator, &[_][]const u8{ "zig-cache", "tmp", tmp.sub_path[0..] });
        break :blk try fs.realpathAlloc(allocator, relative_path);
```